internal/services/discovery/discovery.go

```diff
@@ -0,0 +1,168 @@
+//go:generate mockgen -source $GOFILE -destination ./mock/discovery.go . Service
+package discovery
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/google/uuid"
+	"github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes"
+)
+
+type Service interface {
+	// GetCSPAndRegion discovers the cluster cloud service provider (CSP) and the region the cluster is deployed in by
+	// listing the cluster nodes and inspecting their labels. CSP is retrieved by parsing the Node.Spec.ProviderID property.
+	// Whereas the region is read from the well-known node region labels.
+	GetCSPAndRegion(ctx context.Context) (csp, region *string, reterr error)
+
+	// GetClusterID retrieves the cluster ID by reading the UID of the kube-system namespace.
+	GetClusterID(ctx context.Context) (*uuid.UUID, error)
+
+	// GetKOPSClusterNameAndStateStore discovers the cluster name and kOps state store bucket from the kube-system namespace
+	// annotation. kOps annotates the kube-system namespace with annotations such as this:
+	// * addons.k8s.io/core.addons.k8s.io: '{"version":"1.4.0","channel":"s3://bucket/cluster-name/addons/bootstrap-channel.yaml","manifestHash":"hash"}'
+	// We can retrieve the state store bucket name and the cluster name from the "channel" property of the annotation value.
+	GetKOPSClusterNameAndStateStore(ctx context.Context, log logrus.FieldLogger) (clusterName, stateStore *string, reterr error)
+
+	// GetOpenshiftClusterID discovers the OpenShift cluster ID by reading the cluster ID from the OpenShift cluster version resource.
+	GetOpenshiftClusterID(ctx context.Context) (*string, error)
+
+	// GetOpenshiftClusterName discovers the OpenShift cluster name by reading the cluster name from the OpenShift master machine resource.
+	GetOpenshiftClusterName(ctx context.Context) (*string, error)
```

Comment:
string pointers are a bit useless, can we use empty strings where we don't see a distinction between empty string and null pointer?